### PR TITLE
Rule S3998: Threads should not lock on objects with weak identity

### PR DESCRIPTION
**Pull request by valhristov**
#274

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotLockWeakIdentityObjects.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotLockWeakIdentityObjects : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3998";
+        private const string MessageFormat = "Replace this with a lock against an object with strong identity.";
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotLockWeakIdentityObjects.cs]: I thought you wanted to give clearer explanation about the issue? I was expecting  to see the lock statement in the message :)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotLockWeakIdentityObjects.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var lockStatement = (LockStatementSyntax)c.Node;
+
+                var expressionSymbol = c.SemanticModel.GetSymbolInfo(lockStatement.Expression).Symbol;
+
+                if (expressionSymbol == null ||
+                    !expressionSymbol.GetSymbolType().IsAny(weakIdentityTypes))
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotLockWeakIdentityObjects.cs]: You might want to check for derivative class too, no?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotLockWeakIdentityObjects.cs (hunk 1 above)
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotLockWeakIdentityObjects.cs]: You should have seen the message in RSPEC ;)

Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotLockWeakIdentityObjects.cs]: Does it seems hard for you to display something like:
`Replace 'lock (this)' with a lock against an object with a strong identity.`
If not and you think that's nice, could you make the change?

**Pull request merged**